cmd/promtool/analyze.go

```diff
@@ -0,0 +1,515 @@
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/regexp"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+var (
+	errNotNativeHistogram = fmt.Errorf("not a native histogram")
+	errNotEnoughData      = fmt.Errorf("not enough data")
+
+	sumSuffix = regexp.MustCompile(`_sum\b`)
+)
+
+type QueryAnalyzeConfig struct {
+	metricType string
+	duration   time.Duration
+	time       string
+	matchers   []string
+}
+
+// run retrieves metrics that look like conventional histograms (i.e. have _bucket
+// suffixes) or native histograms, depending on metricType flag.
+func (c *QueryAnalyzeConfig) run(url *url.URL, roundtripper http.RoundTripper) error {
+	if c.metricType != "classichistograms" && c.metricType != "nativehistograms" {
+		return fmt.Errorf("analyze type is %s, must be 'classichistograms' or 'nativehistograms'", c.metricType)
+	}
+
+	ctx := context.Background()
+
+	api, err := newAPI(url, roundtripper, nil)
+	if err != nil {
+		return err
+	}
+
+	var endTime time.Time
+	if c.time != "" {
+		endTime, err = parseTime(c.time)
+		if err != nil {
+			return fmt.Errorf("error parsing time '%s': %w", c.time, err)
+		}
+	} else {
+		endTime = time.Now()
+	}
+
+	if c.metricType == "nativehistograms" {
+		return c.getStatsFromMetrics(ctx, api, endTime, os.Stdout, identity, true, c.matchers)
+	}
+	for _, matcher := range c.matchers {
+		if !sumSuffix.MatchString(matcher) {
+			return fmt.Errorf("every classic histogram matcher must have a '_sum' suffix in the metric name, but the matcher '%s' does not", matcher)
+		}
+	}
+	return c.getStatsFromMetrics(ctx, api, endTime, os.Stdout, replaceSumWithBucket, false, c.matchers)
+}
+
+type transformNameFunc func(name string) string
+
+func identity(name string) string {
+	return name
+}
+
+func replaceSumWithBucket(name string) string {
+	return sumSuffix.ReplaceAllString(name, "_bucket")
+}
+
+func (c *QueryAnalyzeConfig) getStatsFromMetrics(ctx context.Context, api v1.API, endTime time.Time, out io.Writer, transformNameForSeriesSel transformNameFunc, isNative bool, matchers []string) error {
+	metastats := newMetaStatistics()
+	metahistogram := newStatsHistogram()
+	if isNative {
+		for _, matcher := range matchers {
+			seriesSel := seriesSelector(matcher, c.duration)
+			matrix, err := querySamples(ctx, api, seriesSel, endTime)
+			if err != nil {
+				return err
+			}
+			for _, vector := range matrix {
+				stats, err := calcNativeBucketStatistics(vector, metahistogram)
+				if err != nil {
+					if err == errNotNativeHistogram || err == errNotEnoughData {
+						continue
+					}
+					return err
+				}
+				fmt.Fprintf(out, "%s=%v\n", vector.Metric, *stats)
+				metastats.update(stats)
+			}
+		}
+	} else {
+		for _, matcher := range matchers {
+			vector, err := queryLabelSets(ctx, api, matcher, endTime, c.duration)
```

Comment:
Perhaps we could, but isn't it better to break it down into smaller queries in case there are many matching series? Less load on the server at once, and less things this program has to keep in memory while processing.

Reply:
Typical bucket counts are between 2 and 10, so maybe it's not that bad, while the current approach has a lot of possible issues, see my comment above.